src/filter/tree.rs

```diff
@@ -1,37 +1,45 @@
 use super::*;
 
-pub fn dirtree<'a>(
+pub fn pathstree<'a>(
     root: &str,
     input: git2::Oid,
     transaction: &'a cache::Transaction,
 ) -> super::JoshResult<git2::Tree<'a>> {
     let repo = transaction.repo();
-    if let Some(cached) = transaction.get_dir((input, root.to_string())) {
+    if let Some(cached) = transaction.get_paths((input, root.to_string())) {
         return Ok(repo.find_tree(cached)?);
     }
 
     let tree = repo.find_tree(input)?;
     let mut result = tree::empty(&repo);
 
     for entry in tree.iter() {
-        let name = entry.name().ok_or(super::josh_error("INVALID_FILENAME"))?;
-
         if entry.kind() == Some(git2::ObjectType::Blob) {
-            if name == "workspace.josh" {
-                result = replace_child(
-                    &repo,
-                    &std::path::Path::new(
-                        entry.name().ok_or(super::josh_error("no name"))?,
-                    ),
-                    entry.id(),
-                    0o0100644,
-                    &result,
-                )?;
-            }
+            let file_blob = repo.blob(
+                &std::path::Path::new(
+                    &(if root.len() > 0 {
+                        root.to_owned() + "/"
+                    } else {
+                        "".to_string()
+                    } + entry.name().ok_or(super::josh_error("no name"))?),
+                )
+                .to_str()
+                .ok_or(super::josh_error("no name"))?
+                .as_bytes(),
+            )?;
+            result = replace_child(
+                &repo,
+                &std::path::Path::new(
+                    entry.name().ok_or(super::josh_error("no name"))?,
```

Comment:
Also with entry.name() being used multiple times, and requireing error handling, it would make sense to make a local for it.